### ISSUE TITLE
Replace eval with macros

### ISSUE THREAD

**Issue by GiggleLiu** [mroavi/JunctionTrees.jl#12]
The minimum fix should be like 
```julia
using JunctionTrees

problem_number = "34"
problem_filename = joinpath("Promedus_" * problem_number)
problem_dir = joinpath(@__DIR__, "../examples/problems/Promedus/", problem_number)
uai_filepath = joinpath(problem_dir, problem_filename * ".uai")
uai_evid_filepath = joinpath(problem_dir, problem_filename * ".uai.evid")
uai_mar_filepath = joinpath(problem_dir, problem_filename * ".uai.MAR")
td_filepath = joinpath(problem_dir, problem_filename * ".tamaki.td")

reference_marginals = JunctionTrees.read_uai_mar_file(uai_mar_filepath)
obsvars, obsvals = JunctionTrees.read_uai_evid_file(uai_evid_filepath)

# ------------------------------------------------------------------------------
# Posterior marginals given evidence
# ------------------------------------------------------------------------------

macro inferencemodel(uai_filepath, uai_evid_filepath)
    compile_algo(
            eval(uai_filepath),
            uai_evid_filepath = eval(uai_evid_filepath),
        )
end

algo = @inferencemodel uai_filepath uai_evid_filepath

marginals = algo(obsvars, obsvals) |> x -> map(y -> y.vals, x)
```

**Comment by GiggleLiu** [comment]
> Would there be any advantage of using a non-standard string literal over a macro? E.g.:
> 
> ```julia
> algo = compile"problems/asia/asia.uai"
> obsvars, obsvals = Int64[], Int64[]
> marginals = run_algo(algo, obsvars, obsvals)
> ```

It looks good. The problem is how to include also uai_evid_filepath.

**Comment by mroavi** [comment]
Right. You have a point. But putting aside the extra arguments that we might want to pass to `compile_algo`, would this bring benefits over a standard macro?
I think we should go to the approach you propose then. I have a question about macros though. `compile_algo` can take a series of optional parameters that customize the generation of the algorithm. Do macros support optional arguments as well? I couldn't find anything related to this in the docs section about metaprogramming.

**Comment by GiggleLiu** [comment]
> Right. You have a point. But putting aside the extra arguments that we might want to pass to `compile_algo`, would this bring benefits over a standard macro? I think we should go to the approach you propose then. I have a question about macros though. `compile_algo` can take a series of optional parameters that customize the generation of the algorithm. Do macros support optional arguments as well? I couldn't find anything related to this in the docs section about metaprogramming.

It is easy to do, just create a macro like the following
```julia
julia> macro m(ex1, kwargs...)
           :(($(kwargs)...,))
       end
@m (macro with 1 method)

julia> @m a b=3 c=5
(:(b = 3), :(c = 5))
```

**Comment by mroavi** [comment]
Ahh, macros do support `kwargs`. Great. Then I think we should go with this approach. I also want to rename the `algo` function to `infer`.

**Comment by mroavi** [comment]
> The minimum fix should be like
> 
> ```julia
> using JunctionTrees
> 
> problem_number = "34"
> problem_filename = joinpath("Promedus_" * problem_number)
> problem_dir = joinpath(@__DIR__, "../examples/problems/Promedus/", problem_number)
> uai_filepath = joinpath(problem_dir, problem_filename * ".uai")
> uai_evid_filepath = joinpath(problem_dir, problem_filename * ".uai.evid")
> uai_mar_filepath = joinpath(problem_dir, problem_filename * ".uai.MAR")
> td_filepath = joinpath(problem_dir, problem_filename * ".tamaki.td")
> 
> reference_marginals = JunctionTrees.read_uai_mar_file(uai_mar_filepath)
> obsvars, obsvals = JunctionTrees.read_uai_evid_file(uai_evid_filepath)
> 
> # ------------------------------------------------------------------------------
> # Posterior marginals given evidence
> # ------------------------------------------------------------------------------
> 
> macro inferencemodel(uai_filepath, uai_evid_filepath)
>     compile_algo(
>             eval(uai_filepath),
>             uai_evid_filepath = eval(uai_evid_filepath),
>         )
> end
> 
> algo = @inferencemodel uai_filepath uai_evid_filepath
> 
> marginals = algo(obsvars, obsvals) |> x -> map(y -> y.vals, x)
> ```

This approach has an issue with the evaluation of the `uai_filepath` inside the macro. Initially, I thought it was just a macro hygiene problem, but apparently, it is more than that. A macro can only see the expression it's been given, so it (without eval, which is a global construct and doesn't work in e.g. let or @testset blocks) cannot see the values of variables in that expression

One approach is to pass the problem name to the macro as a literal, e.g. "Promedus_14", and then moving the lines
```julia
filepath = joinpath(problem_dir, problem * ".uai")
algo = @mymacro(uai_filepath)
```
inside the macro. However, this would restrict very negatively the interface to construct the posterior marginals algorithms since it will not be able to accept variables with bounded values.
